clients/wallet/src/stellar_wallet.rs

```diff
@@ -25,20 +27,53 @@ pub struct StellarWallet {
 	/// Releasing the lock ensures the sequence number of the account
 	/// has been increased on the network.
 	transaction_submission_lock: Arc<Mutex<()>>,
+	/// Used for caching Stellar transactions before they get submitted.
+	cache: TxEnvelopeStorage,
+}
+
+impl StellarWallet {
+	async fn submit_transaction(
+		&mut self,
+		sequence: SequenceNumber,
+		envelope: TransactionEnvelope,
+		horizon_client: &Client,
+	) -> Result<(TransactionResponse, TransactionEnvelope), Error> {
+		let _ = self.cache.save_to_local(sequence, envelope.clone());
+
+		let client_result = horizon_client
+			.submit_transaction(envelope.clone(), self.is_public_network)
+			.await
+			.map(|response| (response, envelope));
+
+		// for unit testing purposes, file won't be removed immediately.
+		#[cfg(not(test))]
+		let _ = self.cache.remove_transaction(sequence);
```

Comment:
We'd have to check the result of the submission before removing it from the cache. If for example, it's a timeout error, we would need to keep this transaction and retry it. If however, it's eg. a tx_bad_seq error, we can discard the transaction basically. 

Or we change the `resubmit_transactions_from_cache()` method to return errors + the transaction envelope for which the error occurred, so that the transaction envelope is not lost overall, and the caller of the method can decide how to continue.

Reply:
Did you mean Stellar's [timeout](https://developers.stellar.org/api/errors/http-status-codes/horizon-specific/timeout) response? 
I've updated the `HorizonSubmissionError` from 1 param, into a struct with 4 fields. One of the fields is `envelope_xdr`.  The user can deal with it as he like.

Reply:
Sounds good